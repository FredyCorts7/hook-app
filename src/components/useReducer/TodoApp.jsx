import React, { useReducer, useEffect } from 'react';

import { todoReducer } from './todoReducer';

import './styles.css';
import { useForm } from '../../hooks/useForm';

const init = () => {
  return JSON.parse(localStorage.getItem('todos')) || [];

  // return [
  //   {
  //     id: new Date().getTime(),
  //     title: 'Learn react DOM',
  //     description: 'programming with create-react-app',
  //     done: false,
  //   },
  // ];
};

const TodoApp = () => {
  const [todos, dispatch] = useReducer(todoReducer, [], init);

  const [formValues, handleInputChange, reset] = useForm({
    title: '',
    description: '',
  });

  const { title, description } = formValues;

  useEffect(() => {
    localStorage.setItem('todos', JSON.stringify(todos));
  }, [todos]);

  const handleDelete = (todoId) => {
    const action = {
      type: 'delete',
      payload: todoId,
    };

    dispatch(action);
  };

  const handleToggle = (todoId) => {
    const action = {
      type: 'toggle',
      payload: todoId,
    };

    dispatch(action);
  };

  const handleSubmit = (e) => {
    e.preventDefault();

    const newTodo = {
      id: new Date().getTime(),
      title,
      description,
      done: false,
    };

    const action = {
      type: 'add',
      payload: newTodo,
    };

    dispatch(action);
    reset();
  };

  return (
    <div>
      <h1>Todo App {todos.length}</h1>
      <hr />

      <form onSubmit={handleSubmit} className='container'>
        <h3 className='text-center'>Add ToDo</h3>

        <input
          type='text'
          name='title'
          placeholder='ex. learn ...'
          className='form-control'
          autoComplete='off'
          value={title}
          onChange={handleInputChange}
        />

        <input
          type='text'
          name='description'
          placeholder='ex. about that ...'
          className='form-control mt-2'
          autoComplete='off'
          value={description}
          onChange={handleInputChange}
        />

        <button
          type='submit'
          className='btn btn-outline-success btn-block mt-2'
        >
          Add ToDo
        </button>
      </form>

      <ul className='list-group mt-4 container'>
        <h3 className='text-center'>ToDo List</h3>

        {todos.map((todo, index) => (
          <li key={todo.id} className='list-group-item'>
            <div className='row'>
              <p
                className={`${todo.done ? 'complete' : ''} col-8`}
                onClick={() => handleToggle(todo.id)}
              >
                {index + 1}. {todo.title}
              </p>
              <button
                className='btn btn-outline-danger col-4'
                onClick={() => handleDelete(todo.id)}
              >
                Delete
              </button>
            </div>
          </li>
        ))}
      </ul>
    </div>
  );
};

export default TodoApp;
