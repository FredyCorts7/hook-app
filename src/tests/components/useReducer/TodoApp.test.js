import React from 'react';
import { mount, shallow } from 'enzyme';
import TodoApp from '../../../components/useReducer/TodoApp';
import { demoTodos } from '../../fixtures/demoTodos';
import { act } from '@testing-library/react';
import '@testing-library/jest-dom';

describe('Testing to <TodoApp />', () => {
  const wrapper = shallow(<TodoApp />);

  Storage.prototype.setItem = jest.fn(() => {});

  test('should show succesfully', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('should add Todo', () => {
    const wrapper = mount(<TodoApp />);

    act(() => {
      wrapper.find('TodoAdd').prop('handleAddTodo')(demoTodos[0]);
      wrapper.find('TodoAdd').prop('handleAddTodo')(demoTodos[1]);
    });

    expect(wrapper.find('h1').text()).toBe(`Todo App ${2}`);
    expect(localStorage.setItem).toHaveBeenCalledTimes(2);
  });

  test('should delete Todo', () => {
    wrapper.find('TodoAdd').prop('handleAddTodo')(demoTodos[0]);
    wrapper.find('TodoList').prop('handleDelete')(demoTodos[0].id);

    expect(wrapper.find('h1').text()).toBe(`Todo App ${0}`);
  });
});
