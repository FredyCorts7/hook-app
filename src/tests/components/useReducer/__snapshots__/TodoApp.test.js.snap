// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing to <TodoApp /> should show succesfully 1`] = `
<div>
  <h1>
    Todo App 
    0
  </h1>
  <hr />
  <TodoAdd
    handleAddTodo={[Function]}
  />
  <TodoList
    handleDelete={[Function]}
    handleToggle={[Function]}
    todos={Array []}
  />
</div>
`;
